\documentclass[a4paper,12pt]{article}
\usepackage[a4paper,left=25mm,right=20mm,top=20mm,bottom=20mm]{geometry}
\usepackage{t1enc}
\usepackage[utf8]{inputenc}
\usepackage{graphicx}
\usepackage{apacite}
\usepackage{eqlist}
\usepackage{mathtools}
\usepackage{caption}
\usepackage{svg}
\usepackage{amsmath}
\usepackage{float}
\usepackage{indentfirst}
\usepackage{fontspec}
\setmainfont{Times New Roman}
\usepackage{sectsty}
\numberwithin{equation}{section} % how the equation numbers are formed
\newtagform{show_eq}{(Equation.\ }{)}  % how the equation numbers are displayed
\usetagform{show_eq}
\usepackage{caption}
 \captionsetup[figure]{labelfont={bf},name={Fig.},labelsep=period}

\sectionfont{\fontsize{12}{15}\selectfont}

\begin{document}
\begin{center}
	\Large{Patterns of numtogenesis in sixteen different mice strains}
\end{center}
\small{Bálint Biró$^1$, Zoltán Gál$^1$, Michael Brookman$^2$, Orsolya Ivett Hoffmann$^1$}\\ \\
\scriptsize{$^1$ Hungarian University of Agricultural and Life Sciences, Institute of Genetics and Biotechnology, Szent-Györgyi Albert Str. 4, H-2100, Gödöllö, Hungary\\
$^2$ Hanze University of Applied Sciences, Department for Biology and Medical Laboratory Research, Zernikeplein 7, 9747 AS Groningen, Netherlands}

\normalsize
\section{Introduction}
\subsection{Definition and importance of numtogenesis}
\indent The sequences in the nuclear genome with mitochondrial origins are called numts and their integration process itself is called numtogenesis~\shortcite{cat_numt}. \\ \indent The importance of numts were proved in several pathophysiological conditions where oncogene activation, tumor supressor deactivation were the consequences of numt insertion~\shortcite{cancer_numt1,adenocarcinoma_numt,cancer_numt2}. Other than cancer biology, numts have important applications in the fields of phylogenetics~\shortcite{phylogenetic_numt1,phylogenetic_numt2} and forensic research~\shortcite{forensic_numt1,forensic_numt2}. Based on recent studies, numts have an influence on mitochondrial genome editing systems by providing OFF-target cleavage sites in the nuclear genome~\shortcite{off_target_numt}.

\subsection{Molecular background of numtogenesis}
\indent The precondition of numtogenesis is the degradation of the mitochondrial membrane which is due to mitostress inducing factors~\shortcite{radiation_numtogenesis}. These factors could be ionizing radiation, endotoxins, ROS, aging, heat shock etc.. After the membrane damage has happened, the organelle is going to be tagged with a degradation signal which will induce mitophagy. Mitophagy is a special case of autophagy which takes place in the mitophagosome~\shortcite{mitochondrial_degradation}. Mitophagosome is responsible for eliminating and recycling damaged mitochondria~\shortcite{molecular_poltergeists}. During imperfect mitophagy, mitochondrial componenets (i.e. numts) could escape into the cytoplasm. Another explanation is that numts could enter into the cytosol during fusion and fission events~\shortcite{molecular_poltergeists,alt_numtogenesis}. Numts are protected in the cytoplasm against nucleases due to a vacuole mediated process or due to complex formation with histone like proteins. These cytosol located numts can enter the nucleus by membrane fusion~\shortcite{radiation_numtogenesis}. Once the numts are in the nucleus, the Non-Homologous-End-Joining repair mechanism (NHEJ) will insert them into the nuclear genome at double stranded brakes. In the case of NHEJ when no template DNA is available, a nuclease mediated deletion is likely to happen. This will result in a relatively long, single stranded DNA which risks the occurrences of long deletions and translocations. According to alternative explanations, numts serve as template DNAs to prevent bigger mutations~\shortcite{molecular_poltergeists}.

\subsection{Evolutionary background of numtogenesis}
\indent At the very beginning of the eukaryotic evolution, an intracellular symbiotic partnership happened between proteobacteria and Archea~\shortcite{endosymbiotic_origin,mitochondrion_origin}. Throughout this long specialisation process, different intracellular organelles were formed to execute different functions. One of these cellular compartments is mitochondrion which serves as the main machinery for aerobic respiration and participates in many other important cellular processes~\shortcite{mitochondrion_origin}. Mitochondrion also has its own genome which was seriously reduced in size through the coevolution of symbiont and host. This reduction process is called endosymbiotic gene transfer (EGT) and basically it is a genome plasticity phenomenon since parts of mitochondrion's genetic material are integrated into the nuclear genome~\shortcite{EGTheory}. Muller's theory (aka Muller's ratchet) serves as an explanation to EGT which declares that deleterious mutations are going to reduce the size of an asexually isolated genome. Hence in long term the genome will go extinct~\shortcite{mullers_theory,muller_to_mt1,muller_to_mt2}. The uniparental inheritance of the mitochondrial genome asexually isolates it since it cannot be recombined~\shortcite{mt_uniparental_inheritance}. This means that Muller's ratchet is applicable to the mitochondrial genome~\shortcite{muller_to_mt3}. Therefore EGT prevents mtDNA to be totally degraded as it could happen as proposed by Muller by inserting numts into the nuclear genome~\shortcite{EGT_rescues_MT,EGTheory}. Another explanation to EGT is that it is too costly for a cell to maintain a highly complex interactome network of multiple organelles and multiple genomes per each organelle. That is why EGT is functioning~\shortcite{EGT_economy}.

\subsection{Objective}
\indent Mouse is the most commonly used model organism and numerous mice strains serve as model animals for important human diseases~\shortcite{animal_models1,animal_models2,animal_models3,animal_models4}. Numts have been characterized in several species like honeybee~\shortcite{bee_numts}, human~\shortcite{human_numts}, wasps~\shortcite{wasp_numts}, cat~\shortcite{cat_numt} and so on. Numts were also described even in mouse genome~\shortcite{mouse_numts1} but no detailed characterisation were published in terms of different mice strains. Therefore our primary goal was to explore the patterns of numts in different mice strains and to see whether they are different from the numts in \textit{Mus musculus musculus}.

\section{Materials and Methods}
\indent Nuclear and mitochondrial DNAs were acquired from Ensembl and NCBI databases~\shortcite{ncbi,ensembl}. The mitochondrion of \textit{Mus musculus musculus} was annotated using MITOS server hosted by the University of Leipzig~\shortcite{mitos}. For the numt comparisons 12 inbred strains (129S1/SvImJ, A/J, AKR/J, BALB/cJ, C3H/HeJ, C57BL/6NJ, CBA/J, DBA/2J, FVB/NJ, LP/J, NZO/HlLtJ, and NOD/ShiLtJ) and four "wild derived" strains (CAST/EiJ, PWK/PhJ, WSB/EiJ, SPRET/EiJ) were selected based on a previously published genome comparison analysis~\shortcite{mice_strains}. Double mtDNA and gDNA were aligned using LASTAL (v1219)~\shortcite{lastal} with the scoring scheme of + 1 for matches, −1 for mismatches, 7 for gap-open penalty and 1 for gap-extension penalty. Using doubled mtDNA, we were able to identify numts that are located on the linearization point of the mtDNA. After that, alignments were filtered based on their e-values as proposed by Tsuji~\shortcite{e_threshold}.  False positive alignments that were the results of using double mtDNA were also discarded.\\ \indent To examine normality, Anderson-Darling test was performed at 0.05 p value. t-test or Wilcoxon signed rank test was performed based on the result of normality testing. All the statistical calculations were conducted in Scipy (v1.6.2)~\shortcite{scipy}. Whether tRNA genes are expressed were derived from their structures using free energy values which were calculated with seqfold~\shortcite{seqfold} based on a previous research paper~\shortcite{trna_expression}. \\ Pairwise divergences were calculated using the modified Kimura 2 parameter~\shortcite{kimura2} which tolerates gaps in the alignments. The modified Kimura 2 formula is described by Equation~\ref{eq:kimura2}.

\begin{equation}\label{eq:kimura2}
	K=\frac{3}{4}w\log w- \frac{w}{2} \log (S-P) \sqrt{S+P-Q}
\end{equation}\\
where $w$=the probability that the given position contains a nucleotide,\\
$S$=$n_1$/$n$,\\
$n_1$=the number of positions where the two aligned sequences contain the same nucleotide,\\
$n$=total number of nucleotides,\\
$P$=$n_2$/$n$,\\
$n_2$=number of transition type mutations,\\
$n_3$=number of transversion type mutations.\\

Pairwise divergence values were calculated using a sliding window approach with 1kb window size and 10bp step size.\\ \indent For the free energy calculations, the corresponding tRNA sequences for each strains were acquired with SAMTOOLS's (v1.6) faidx function~\shortcite{samtools}.\\ \indent The phylogenetic analysis was conducted in R with phangorn~\shortcite{phangorn}. The Maximum-Parsimony tree was constructed using Jukes-Cantor distance with 100 bootstrap cycles. For tree construction cytochrome b (\textit{CYTB}) and corresponding numt sequences were used as described by a previous study~\shortcite{phylogenetic_method} with rat as outgroup.\\ \indent All figures were created in matplotlib (v3.4.3) and seaborn (v0.11.2)~\shortcite{matplotlib,seaborn}.

\section{Results}
\indent 152 numts can be identified in the \textit{Mus musculus musculus} genome. There is a huge variability in terms of numt numbers on the chromosomes. For example, chr1 contains 14 numts, while no numt can be found on chr19. The numts cover the total mitochondrion. The longest numt (4654 bp) covers 6 protein coding genes~(Fig.\ref{fig:mm_numts_corr}/a). There is a strong correlation (Pearson correlation coefficients: 0.77) between the chromosome size and the number of numts on a given chromosome~(Fig.\ref{fig:mm_numts_corr}/b). The shortest chromosome contains the smallest number of numts while the longest chromosome contains the highest number of numts. However this relationship is not true for any of the strains investigated.
\begin{figure}[H]
    \centering
    \captionsetup{justification=centering}
    \includegraphics[width=.8\textwidth]{../results/mm_numts_corr.png}
    \caption{Patterns of  \textit{Mus musculus musculus} numts. Distribution of  \textit{Mus musculus musculus} with the genomic locations (a) and the correlation between the number of numts and the size of the corresponding chromosome (b). Small tRNA genes are not part of the annotation.}
    \label{fig:mm_numts_corr}
\end{figure}

\indent When investigating the distribution of numts along the mitochondria in different strains, it turns out that the majority of the numts are the same as in the case of  \textit{Mus musculus musculus}. However exceptions do exist. For example as we have already described above,  numts originate from the whole mitochondrion in case of  \textit{Mus musculus musculus} but not in the case of several inbred and wild derived strains. Differences are also present in the lengths of the numts. For instance, the longest numt in the \textit{Mus musculus musculus} genome which locates on chr1 is missing from all the strains investigated~(Fig.\ref{fig:mitochondrial_origins}.). 
\begin{figure}[H]
    \centering
    \captionsetup{justification=centering}
    \includegraphics[width=.8\textwidth]{../results/mitochondrial_origins.png}
    \caption{Mitochondrial origin of numts in different mice strains. The upper row contains the wild derived strains. Blue arrows indicate mitochondrial regions where numts do not originate from.}
    \label{fig:mitochondrial_origins}
\end{figure}

\indent There are common patterns in terms of  numt coverage. Interestingly the regions of mitochondria that contain the linearization point (the start and the end of the linearized mitochondrion) are highly covered by numts. This region contains the \textit{tRNA-S} coding gene and the D-loop. Another numt dense region which is present in every strains, covers \textit{atp6}, \textit{cox3}, \textit{nad3} and \textit{nad4l} genes. There are two general numt sparse regions. The first one partially covers \textit{tRNA-S} and \textit{tRNA-L} genes while the second one covers \textit{nad4}~(Fig.\ref{fig:mt_sliding_windows}.).\\ \indent From the wild derived strains casteij, spreteij and pwkphj while from inbred strains aj and akrj are clustered together when it comes to nuclotides along mitochondria involved in numtogenesis. At the same time wsbeij and the rest of the inbred strains are also clustered together~(Fig.\ref{fig:numts_bp_corr}.).
\begin{figure}[H]
    \centering
    \captionsetup{justification=centering}
    \includegraphics[width=.8\textwidth]{../results/mt_sliding_windows.png}
    \caption{Numt contents along mitochondria. Colored areas show common patterns. Green coloring is corresponding for dense numt region while red coloring is corresponding for numt sparse region.}
    \label{fig:mt_sliding_windows}
\end{figure}
\begin{figure}[H]
    \centering
    \captionsetup{justification=centering}
    \includegraphics[width=.5\textwidth]{../results/numts_bp_corr.png}
    \caption{Clustered heatmap representation of the Pearson correlation matrix of nuclotides involved in numtogenesis along mitochondria.}
    \label{fig:numts_bp_corr}
\end{figure}

\indent The wild derived strains casteij, spreteij and pwkphj show higher pairwise divergence values when total mitochondria are compared with the mitochondrion of  \textit{Mus musculus musculus}. Surprisingly, the fourth wild derived strain wsbeij does not differ significantly from \textit{Mus musculus musculus} while the inbred nzohiltj strain shows elevated pairvise distance values to some extent~(Fig.\ref{fig:modK2s}.). \\ \indent Maxmimum-Parsimony phylogenetic tree supports two monophyletic clades, namely the mitochondrial CYTB and the corresponding numt sequences. The result of the phylogenetic analysis resembles the whole mitochondria divergence values even though the sequences of only one gene were used. In the CYTB clade the wild derived strains casteij, spreteij and pwkphj plus the inbred nzohiltj strain differ from the other strains. However nzohiltj is very close to the other inbred strains. In the numt clade the wild derived strains casteij, spreteij and pwkphj are different than the rest of the strains. In this clade the inbred strain nzohiltj is clustered together with the inbred strains unlike in the case of CYTB clade, pairwise divergence and numt coverage analysis. Intriguingly wsbeij is clustered together with the inbred strains in both clades even though it is a wild dervied strain~(Fig.\ref{fig:bs_mp}.).
\begin{figure}[H]
    \centering
    \captionsetup{justification=centering}
    \includegraphics[width=.8\textwidth]{../results/modK2s.png}
    \caption{Pairwise divergence values. The upper row contains the wild derived strains.}
    \label{fig:modK2s}
\end{figure}

\begin{figure}[H]
    \centering
    \captionsetup{justification=centering}
    \includegraphics[width=.8\textwidth]{../results/bs_mp.png}
    \caption{Maximum-Parsimony tree based on Jukes-Cantor distance with 100 bootstrap and rat CYTB as outgroup.}
    \label{fig:bs_mp}
\end{figure}

\indent During the analysis of free energy values of the folding of tRNA sequences and their corresponding numts no case was shown where numt folding's \Delta G value was the same as tRNA folding's. In most of the cases tRNA folding's \Delta G was smaller than numt folding's \Delta G~(Fig.\ref{fig:deltaGs}.).
\begin{figure}[H]
    \centering
    \captionsetup{justification=centering}
    \includegraphics[width=.8\textwidth]{../results/connected_deltaGs.png}
    \caption{The free energy values of the predicted structures  of tRNAs and their corresponding numts. The upper row contains the wild derived strains.}
    \label{fig:deltaGs}
\end{figure}

\section{Conclusions}
\indent Like in many other eukaryotic organisms~\shortcite{molecular_poltergeists,mouse_numts1,e_threshold,wasp_numts}, numts are also present in the genome of  \textit{Mus musculus musculus}. However the different patterns of numts in mice strains have not been described previously. Hence in this study numts of  divergent mice strains were investigated.\\ \indent We described 152 numts in the \textit{Mus musculus musculus} genome which is comparable with previous studies~\shortcite{mouse_numts1,e_threshold}.\\ \indent There is a strong correlation in the \textit{Mus musculus musculus} genome between chromosomal length and the number of the numts on the given chromosome. This correlation also exists in the human genome~\shortcite{chrsize_corr_numtnumber,human_numts}. In general, there is a higher gene density in case of shorter chromosomes. Since natural selection tries to avoid insertional mutagenesis (eg. intragenic numts), there is a higher possibility for less gene dense, longer chromosomes to tolerate a numt integration without serious consequences~\shortcite{chrsize_corr_numtnumber}. However this is not the case in any of the strains investigated. No correlation was found between the above mentioned attributes also in the case of honey bee~\shortcite{bee_numts}. \\ \indent The majority of \textit{Mus musculus musculus} numts were the same in the strains investigated. However the longest numt (4654 bp on chr1) could not be detected in any of the strains. This situation was found in a couple of other organisms and it is originated from a fragmentation event after the given numt has been integrated into the nuclear genome~\shortcite{bee_numts,numt_fragmentation}.\\ \indent The investigation of numt coverage along the mitochondria reveals that mitochondrial nucleotides are present in several copies. Numt coverage results also proved that the representation of the nucleotides along the mitochondria differ. Hence that over- as well as under-represented regions do exist. This kind of imbalanced representation of mitochondrial nucleotides were reported in several mammalian species~\shortcite{human_numts,e_threshold}.\\ \indent All of our experiments that somehow integrate data about the similarity of numts or mitochondria (numt content along mitochondria, pairwise divergence, phylogeny) resemble the relationship between \textit{Mus musculus musculus} and spreteij as to our current knowledge. Namely that from the strains investigated, spreteij seems to be the most distantly related compared to \textit{Mus musculus musculus}. The phenomenon that spreteij is the most distant strain from the strains examined also applies to when total nuclear genomes are being compared~\shortcite{mice_strains}.\\ \indent Since the mitochondrial genetic code and the nuclear genetic code are different~\shortcite{genetic_codes}, the same DNA sequence would be resulted in different RNA molecules depending on the used genetic code. In addition, the nucleotide composition of numts are adjusted to the mitochondrial machinery. Hence numts are considered to be pseudogenes that are unable to code the genetic information that they used to code. However there is still a chance for numts to code functional tRNAs since tRNAS coded by the nucleus and tRNAs coded by mitochondrion would be identical~\shortcite{genomic_origin_of_mtrnas}. Although we found no evidence for expressing numts that are corresponding to tRNA genes. All the tRNA numts were altered in sequence and the free energy of the folding of these numts were always higher than the free energy of the folding of corresponding tRNA genes. However the free energy values of the folding of numts were still negative, which means that the folding itself is a spontaneous process. But tRNA genes had the minimum free energy which means the most stable structure from a thermodynamic point of view~\shortcite{rna_structure}. Our results regarding tRNA expression are in agreement with the results in the case of cattle~\shortcite{trna_expression}.\\ \indent In this paper we described the patterns of numts in different mice strains and characterised the differences between each strains. Our results can contribute to distinct phylogenetic and disease modell studies.

\section{Acknowledgements}
\indent B.B. received NTP-NFTÖ-21-B-0127 scholarship. O.I.H. was funded by NKFIH OTKA FK 124708, János Bolyai Research Scholarship of the Hungarian Academy of Sciences and New National Excellence Program of the Ministry for Innovation and Technology from the source of the National Research, Development and Innovation Fund (ÚNKP-21-5).
\section{References}
\renewcommand{\refname}{}
\bibliographystyle{apacite}
\bibliography{references.bib}


\end{document}